File: c/src/game/game.c
#include "game.h"
#include <stdlib.h>

static int damage_player(Entity *e, int amount) {
    Player *p = (Player *)e->ctx;
    if (p->health <= amount) {
        p->health = 0;
        e->alive = false;
    } else {
        p->health -= amount;
    }
    p->damaged = true;
    e->color = ECOLOR_NONE;
    return 0;
}

static void update_player(Entity *e) {
    Player *p = (Player *)e->ctx;
    p->ticks++;
    if (p->damaged) {
        p->damaged = false;
    } else {
        e->color = ECOLOR_CYAN;
    }
}

static int damage_enemy(Entity *e, int amount) {
    Enemy *p = (Enemy *)e->ctx;
    if (p->health <= amount) {
        p->health = 0;
        e->alive = false;
        return p->score;
    }
    p->health -= amount;
    return 0;
}

static void update_enemy(Entity *e) {
    Enemy *p = (Enemy *)e->ctx;

    e->color = ECOLOR_NONE;
    p->last_fired += 1;

    if (p->last_fired >= (p->fire_freq - 1)) {
        e->color = ECOLOR_RED;
    }

    if (p->last_fired >= p->fire_freq) {
        Entity *b = game_spawn_bullet(p->game, e->x, e->y + 1, p->damage, ENTITY_ENEMY);
        b->shape = "|";
        p->last_fired = 0;
    }
}

static void update_bullet(Entity *e) {
    Bullet *b = (Bullet *)e->ctx;
    b->last_moved++;

    if (b->last_moved < b->move_freq) {
        return;
    }
    b->last_moved = 0;

    if (b->owner == ENTITY_ENEMY) {
        if (entity_collision(e, &b->game->player)) {
            entity_damage(&b->game->player, b->damage);
            e->alive = false;
        }
        // Move down
        if (e->y + 1 >= b->game->bounds.h) {
            e->alive = false;
            return;
        }
        e->y += 1;
    } else {
        for (int i = 0; i < b->game->enemies.count; i++) {
            if (entity_collision(e, &b->game->enemies.items[i])) {
                int score = entity_damage(&b->game->enemies.items[i], b->damage);
                b->game->score += score;
                b->game->shots_hit++;
                if (score > 0) {
                    b->game->kills++;
                }
                e->alive = false;
                break;
            }
        }
        // Move up
        if (e->y - 1 < 0) {
            e->alive = false;
            return;
        }
        e->y -= 1;
    }
}

Entity *game_spawn_player(Game *g, int x, int y, int health, int damage, int cooldown) {
    Player *p = (Player *)malloc(sizeof(Player));
    p->health = health;
    p->damage = damage;
    p->cooldown = cooldown;
    p->ticks = cooldown;
    p->damaged = false;

    Entity e = {
        .x = x,
        .y = y,
        .alive = true,
        .shape = "A",
        .type = ENTITY_PLAYER,
        .damage = damage_player,
        .update = update_player,
        .color = ECOLOR_CYAN,
        .ctx = p,
    };

    g->player = e;
    return &g->player;
}

Entity *game_spawn_enemy(Game *g, int x, int y, int health, int score, int fire_freq, int damage) {
    Enemy *p = (Enemy *)malloc(sizeof(Enemy));
    p->health = health;
    p->score = score;
    p->game = g;
    p->fire_freq = fire_freq;
    p->last_fired = 0;
    p->damage = damage;

    Entity e = {
        .x = x,
        .y = y,
        .alive = true,
        .shape = "V",
        .type = ENTITY_ENEMY,
        .damage = damage_enemy,
        .update = update_enemy,
        .color = ECOLOR_NONE,
        .ctx = p,
    };

    list_append(&g->enemies, e);
    return &g->enemies.items[g->enemies.count - 1];
}

Entity *game_spawn_bullet(Game *g, int x, int y, int damage, EntityType owner) {
    Bullet *b = (Bullet *)malloc(sizeof(Bullet));
    b->damage = damage;
    b->owner = owner;
    b->game = g;
    b->move_freq = 1;
    b->last_moved = 0;

    Entity e = {
        .x = x,
        .y = y,
        .alive = true,
        .shape = "0",
        .type = ENTITY_BULLET,
        .damage = NULL,
        .update = update_bullet,
        .color = ECOLOR_NONE,
        .ctx = b,
    };

    list_append(&g->bullets, e);
    return &g->bullets.items[g->bullets.count - 1];
}

void game_init(Game *g, int w, int h, i64 tps) {
    Game gi = {
        .bounds = {.w = w, .h = h},
        .level = 1,
        .score = 0,
        .status = GAME_STATUS_RUNNING,
        .enemies = {0},
        .bullets = {0},
        .player = {0},
        .last_update = 0,
        .tps = tps,
        .grid = NULL,
        .shots_fired = 0,
        .shots_hit = 0,
        .kills = 0,
        .elapsed_time = 0,
        .time_seconds = 0,
        .accuracy_percent = 0,
    };
    *g = gi;
    game_spawn_player(g, (w - 1) / 2, h - 1, 5, 1, 2);
}

void game_free(Game *g) {
    if (g == NULL) return;

    for (int i = 0; i < g->bullets.count; i++) {
        free(g->bullets.items[i].ctx);
    }
    list_free(g->bullets);

    for (int i = 0; i < g->enemies.count; i++) {
        free(g->enemies.items[i].ctx);
    }
    list_free(g->enemies);

    free(g->player.ctx);

    if (g->grid) {
        grid_free(g->grid);
    }
}

void game_reset(Game *g) {
    for (int i = 0; i < g->bullets.count; i++) {
        free(g->bullets.items[i].ctx);
    }
    g->bullets.count = 0;

    for (int i = 0; i < g->enemies.count; i++) {
        free(g->enemies.items[i].ctx);
    }
    g->enemies.count = 0;

    // Reset player
    Player *p = (Player *)g->player.ctx;
    p->health = 5;
    p->ticks = p->cooldown;
    p->damaged = false;
    g->player.alive = true;
    g->player.x = (g->bounds.w - 1) / 2;
    g->player.y = g->bounds.h - 1;
    g->player.color = ECOLOR_CYAN;

    // Reset game state
    g->level = 1;
    g->score = 0;
    g->status = GAME_STATUS_RUNNING;

    // Reset statistics
    g->shots_fired = 0;
    g->shots_hit = 0;
    g->kills = 0;
    g->elapsed_time = 0;
    g->time_seconds = 0;
    g->accuracy_percent = 0;
}

void game_update(Game *g, i64 udt) {
    g->elapsed_time += udt;

    g->time_seconds = (int)(g->elapsed_time / US_PER_SEC);
    if (g->shots_fired > 0) {
        g->accuracy_percent = (g->shots_hit * 100) / g->shots_fired;
    } else {
        g->accuracy_percent = 0;
    }

    for (int i = 0; i < g->bullets.count; i++) {
        entity_update(&g->bullets.items[i]);
    }

    for (int i = 0; i < g->enemies.count; i++) {
        entity_update(&g->enemies.items[i]);
    }

    entity_update(&g->player);
}

void game_remove_dead_entities(Game *g) {
    int i = 0;
    while (i < g->bullets.count) {
        if (!g->bullets.items[i].alive) {
            free(g->bullets.items[i].ctx);
            list_remove_unordered(&g->bullets, i);
            continue;
        }
        i++;
    }

    i = 0;
    while (i < g->enemies.count) {
        if (!g->enemies.items[i].alive) {
            free(g->enemies.items[i].ctx);
            list_remove_unordered(&g->enemies, i);
            continue;
        }
        i++;
    }

    if (!g->player.alive) {
        g->status = GAME_STATUS_GAME_OVER;
    }
}

void game_process_input(Game *g, Event *ev) {
    if (ev->type != EVENT_KEY) return;

    Player *p = (Player *)g->player.ctx;

    // Movement
    if (key_is_char(ev, 'a') || key_is_char(ev, 'A') || key_is(ev, KEY_LEFT)) {
        if (g->player.x > 0) {
            g->player.x--;
        }
    } else if (key_is_char(ev, 'd') || key_is_char(ev, 'D') || key_is(ev, KEY_RIGHT)) {
        if (g->player.x < g->bounds.w - 1) {
            g->player.x++;
        }
    }
    // Fire
    else if (key_is_char(ev, ' ')) {
        if (p->ticks >= p->cooldown) {
            game_spawn_bullet(g, g->player.x, g->player.y - 1, p->damage, ENTITY_PLAYER);
            p->ticks = 0;
            g->shots_fired++;
        }
    }
}

// Cell draw callback for entities
static void draw_entity_cell(void *ctx, Screen *s, int x, int y) {
    Entity *e = (Entity *)ctx;
    entity_draw(e, s, x, y);
}

void game_draw(void *ctx, Screen *s, BBox b) {
    Game *g = (Game *)ctx;

    if (g->grid == NULL) {
        g->grid = grid_new(g->bounds.w, g->bounds.h);
    }

    for (u32 y = 0; y < g->grid->rows; y++) {
        for (u32 x = 0; x < g->grid->cols; x++) {
            grid_at(g->grid, x, y)->draw = NULL;
        }
    }

    for (int i = 0; i < g->bullets.count; i++) {
        Entity *e = &g->bullets.items[i];
        GridCell *c = grid_at(g->grid, e->x, e->y);
        if (c) {
            c->ctx = e;
            c->draw = draw_entity_cell;
        }
    }

    for (int i = 0; i < g->enemies.count; i++) {
        Entity *e = &g->enemies.items[i];
        GridCell *c = grid_at(g->grid, e->x, e->y);
        if (c) {
            c->ctx = e;
            c->draw = draw_entity_cell;
        }
    }

    Entity *e = &g->player;
    if (e->alive) {
        GridCell *c = grid_at(g->grid, e->x, e->y);
        if (c) {
            c->ctx = e;
            c->draw = draw_entity_cell;
        }
    }

    grid_draw(g->grid, s, b);
}

File: rust/src/game/game.rs
//! Game state container

use crate::common::US_PER_SEC;
use crate::game::entity::{Entity, EntityColor, EntityCommand, Owner};
use crate::input::{self, Event, KeyCode};
use crate::tui::screen::Screen;
use crate::ui::widget::BBox;

/// Game status
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum GameStatus {
    Running,
    Paused,
    Finished,
    GameOver,
}

/// Game bounds
#[derive(Clone, Copy, Debug)]
pub struct Bounds {
    pub w: i32,
    pub h: i32,
}

/// Game state
pub struct Game {
    bounds: Bounds,
    level: i32,
    score: i32,
    status: GameStatus,

    player: Option<Entity>,
    enemies: Vec<Entity>,
    bullets: Vec<Entity>,

    tps: i64,

    // Statistics
    shots_fired: i32,
    shots_hit: i32,
    kills: i32,
    elapsed_time: i64,
}

#[allow(unused)]
impl Game {
    pub fn new(w: i32, h: i32, tps: i64) -> Self {
        let mut game = Game {
            bounds: Bounds { w, h },
            level: 1,
            score: 0,
            status: GameStatus::Running,
            player: None,
            enemies: Vec::new(),
            bullets: Vec::new(),
            tps,
            shots_fired: 0,
            shots_hit: 0,
            kills: 0,
            elapsed_time: 0,
        };

        game.spawn_player((w - 1) / 2, h - 1, 5, 1, 2);
        game
    }

    pub fn bounds(&self) -> Bounds {
        self.bounds
    }
    pub fn level(&self) -> i32 {
        self.level
    }
    pub fn score(&self) -> i32 {
        self.score
    }
    pub fn status(&self) -> GameStatus {
        self.status
    }
    pub fn tps(&self) -> i64 {
        self.tps
    }
    #[allow(dead_code)]
    pub fn shots_fired(&self) -> i32 {
        self.shots_fired
    }
    #[allow(dead_code)]
    pub fn shots_hit(&self) -> i32 {
        self.shots_hit
    }
    pub fn kills(&self) -> i32 {
        self.kills
    }
    pub fn time_seconds(&self) -> i32 {
        (self.elapsed_time / US_PER_SEC) as i32
    }
    pub fn accuracy_percent(&self) -> i32 {
        if self.shots_fired > 0 {
            (self.shots_hit * 100) / self.shots_fired
        } else {
            0
        }
    }
    pub fn bullet_count(&self) -> usize {
        self.bullets.len()
    }
    pub fn enemy_count(&self) -> usize {
        self.enemies.len()
    }

    pub fn player(&self) -> Option<&Entity> {
        self.player.as_ref()
    }
    pub fn player_health(&self) -> i32 {
        self.player
            .as_ref()
            .and_then(|p| p.as_player())
            .map(|d| d.health)
            .unwrap_or(0)
    }

    pub fn set_status(&mut self, status: GameStatus) {
        self.status = status;
    }
    pub fn increment_level(&mut self) {
        self.level += 1;
    }

    pub fn spawn_player(&mut self, x: i32, y: i32, health: i32, damage: i32, cooldown: i32) {
        self.player = Some(Entity::new_player(x, y, health, damage, cooldown));
    }

    pub fn spawn_enemy(
        &mut self,
        x: i32,
        y: i32,
        health: i32,
        score: i32,
        fire_freq: i32,
        damage: i32,
    ) {
        self.enemies
            .push(Entity::new_enemy(x, y, health, score, fire_freq, damage));
    }

    pub fn spawn_bullet(&mut self, x: i32, y: i32, damage: i32, owner: Owner) {
        self.bullets.push(Entity::new_bullet(x, y, damage, owner));
    }

    pub fn reset(&mut self) {
        self.enemies.clear();
        self.bullets.clear();

        if let Some(player) = &mut self.player {
            if let Entity::Player { data, .. } = player {
                data.health = 5;
                data.ticks = data.cooldown;
                data.damaged = false;
            }
            let base = player.base_mut();
            base.alive = true;
            base.x = (self.bounds.w - 1) / 2;
            base.y = self.bounds.h - 1;
            base.color = EntityColor::Cyan;
        }

        self.level = 1;
        self.score = 0;
        self.status = GameStatus::Running;
        self.shots_fired = 0;
        self.shots_hit = 0;
        self.kills = 0;
        self.elapsed_time = 0;
    }

    pub fn update(&mut self, delta_us: i64) {
        self.elapsed_time += delta_us;

        let mut commands = Vec::new();

        if let Some(player) = &mut self.player {
            Self::update_player(player);
        }

        for enemy in &mut self.enemies {
            if let Some(cmd) = Self::update_enemy(enemy) {
                commands.push(cmd);
            }
        }

        let player_pos = self
            .player
            .as_ref()
            .filter(|p| p.is_alive())
            .map(|p| (p.x(), p.y()));

        let enemy_positions: Vec<(i32, i32)> = self
            .enemies
            .iter()
            .map(|e| (e.x(), e.y()))
            .collect();

        for bullet in &mut self.bullets {
            if let Some(cmds) =
                Self::update_bullet(bullet, self.bounds, player_pos, &enemy_positions)
            {
                commands.extend(cmds);
            }
        }

        for cmd in commands {
            match cmd {
                EntityCommand::SpawnBullet {
                    x,
                    y,
                    damage,
                    owner,
                } => {
                    self.spawn_bullet(x, y, damage, owner);
                }
                EntityCommand::DamagePlayer(amount) => {
                    if let Some(player) = &mut self.player {
                        player.damage(amount);
                    }
                }
                EntityCommand::DamageEnemy { index, amount } => {
                    if index < self.enemies.len() {
                        let score = self.enemies[index].damage(amount);
                        self.score += score;
                        self.shots_hit += 1;
                        if score > 0 {
                            self.kills += 1;
                        }
                    }
                }
            }
        }
    }

    fn update_player(player: &mut Entity) {
        if let Entity::Player { base, data } = player {
            data.ticks += 1;
            if data.damaged {
                data.damaged = false;
            } else {
                base.color = EntityColor::Cyan;
            }
        }
    }

    fn update_enemy(enemy: &mut Entity) -> Option<EntityCommand> {
        if let Entity::Enemy { base, data } = enemy {
            base.color = EntityColor::None;
            data.last_fired += 1;

            if data.last_fired >= (data.fire_freq - 1) {
                base.color = EntityColor::Red;
            }

            if data.last_fired >= data.fire_freq {
                let cmd = EntityCommand::SpawnBullet {
                    x: base.x,
                    y: base.y + 1,
                    damage: data.damage,
                    owner: Owner::Enemy,
                };
                data.last_fired = 0;
                return Some(cmd);
            }
        }
        None
    }

    fn update_bullet(
        bullet: &mut Entity,
        bounds: Bounds,
        player_pos: Option<(i32, i32)>,
        enemy_positions: &[(i32, i32)],
    ) -> Option<Vec<EntityCommand>> {
        let Entity::Bullet { base, data } = bullet else {
            return None;
        };

        data.last_moved += 1;

        if data.last_moved < data.move_freq {
            return None;
        }
        data.last_moved = 0;

        let mut commands = Vec::new();

        if data.owner == Owner::Enemy {
            if let Some((px, py)) = player_pos {
                if base.x == px && base.y == py {
                    commands.push(EntityCommand::DamagePlayer(data.damage));
                    base.alive = false;
                    return Some(commands);
                }
            }

            // Move down
            if base.y + 1 >= bounds.h {
                base.alive = false;
            } else {
                base.y += 1;
            }
        } else {
            for (i, &(ex, ey)) in enemy_positions.iter().enumerate() {
                if base.x == ex && base.y == ey {
                    commands.push(EntityCommand::DamageEnemy {
                        index: i,
                        amount: data.damage,
                    });
                    base.alive = false;
                    return Some(commands);
                }
            }

            // Move up
            if base.y - 1 < 0 {
                base.alive = false;
            } else {
                base.y -= 1;
            }
        }

        Some(commands)
    }

    pub fn remove_dead_entities(&mut self) {
        self.bullets.retain(|b| b.is_alive());
        self.enemies.retain(|e| e.is_alive());

        if let Some(player) = &self.player {
            if !player.is_alive() {
                self.status = GameStatus::GameOver;
            }
        }
    }

    pub fn process_input(&mut self, event: &Event) {
        let Event::Key(key) = event else { return };

        let Some(player) = &mut self.player else {
            return;
        };

        let px = player.x();
        let py = player.y();

        if input::is_char(key, 'a') || input::is_key(key, KeyCode::Left) {
            if px > 0 {
                player.set_x(px - 1);
            }
        } else if input::is_char(key, 'd') || input::is_key(key, KeyCode::Right) {
            if px < self.bounds.w - 1 {
                player.set_x(px + 1);
            }
        } else if input::is_char(key, ' ') {
            if let Entity::Player { data, .. } = player {
                if data.ticks >= data.cooldown {
                    let damage = data.damage;
                    data.ticks = 0;
                    self.shots_fired += 1;
                    self.bullets
                        .push(Entity::new_bullet(px, py - 1, damage, Owner::Player));
                }
            }
        }
    }

    pub fn draw(&self, screen: &mut Screen, bbox: BBox) {
        let x_pad = (bbox.w as i32 / self.bounds.w) / 2;
        let y_pad = (bbox.h as i32 / self.bounds.h) / 2;

        for bullet in &self.bullets {
            let xx =
                (bullet.x() * bbox.w as i32 / self.bounds.w + bbox.x as i32 + x_pad) as usize;
            let yy =
                (bullet.y() * bbox.h as i32 / self.bounds.h + bbox.y as i32 + y_pad) as usize;
            bullet.draw(screen, xx, yy);
        }

        for enemy in &self.enemies {
            let xx = (enemy.x() * bbox.w as i32 / self.bounds.w + bbox.x as i32 + x_pad) as usize;
            let yy = (enemy.y() * bbox.h as i32 / self.bounds.h + bbox.y as i32 + y_pad) as usize;
            enemy.draw(screen, xx, yy);
        }

        if let Some(player) = &self.player {
            if player.is_alive() {
                let xx =
                    (player.x() * bbox.w as i32 / self.bounds.w + bbox.x as i32 + x_pad) as usize;
                let yy =
                    (player.y() * bbox.h as i32 / self.bounds.h + bbox.y as i32 + y_pad) as usize;
                player.draw(screen, xx, yy);
            }
        }
    }
}
